Implement PCA9420 PMIC Support into Zephyr

danieldegrasse opened the pull request:
This PR adds support for the NXP PCA9420 PMIC into Zephyr. This PMIC is present on the NXP MIMXRT685 EVK.

The PMIC has 4 regulators, and supports variable voltage and current limit control via I2C. This driver supports the generic
regulator driver to allow disabling and reenabling each regulator on the PMIC, but it also implements an extended regulator API
that allows setting current and voltage limits. This API is defined in 'include/regulators/consumer.h' and is based on the [Linux regulator api](https://github.com/torvalds/linux/blob/master/include/linux/regulator/consumer.h), but uses the devicetree to define configuration registers and voltage levels, while Linux does this in code.

This PR also enables the I2C peripheral FLEXCOMM15 on the RT685, as this was required to test the PMIC present on the board.

danieldegrasse committed:
dts: rt685: enabled flexcomm15

the RT685 contains an additional flexcomm peripheral, that supports
only I2C. This commit adds this peripheral to the device tree,
and enables pins and clocks for flexcomm15.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: regulator: add i2c regulator driver

This commit adds a generic i2c regulator driver, and enables the NXP
PCA9420 PMIC IC using this driver. The regulator driver also exposes an
additional API in include/drivers/regulator/consumer.h, which allows
drivers to implement support for adjusting voltage levels and current
limits, if their device supports it.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: pca9420: Added voltage level support for PCA9420

This commit adds support for editing the voltage levels on the buck
regulators on the PCA9420 PMIC

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse committed:
drivers: pca9420: added support for current limit setting on PCA9420

Added ability to set VIN current limit when using the pca9420 PMIC

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

mbolivar-nordic commented on boards/arm/mimxrt685_evk/mimxrt685_evk_cm33.dts:
```diff
@@ -203,6 +204,79 @@ i2s1: &flexcomm3 {
+
+		pca9420_sw1: sw1_buck {
+			label = "SW1_BUCK_REG";
+			compatible = "regulator-i2c";
```
I've never seen things done this way for an I2C device. Do you have any prior examples elsewhere, e.g. in a Linux binding? I was expecting to see a binding for the PCA9420 itself somewhere.

danieldegrasse replied:
Here's how it is done in the Linux Kernel: https://github.com/torvalds/linux/blob/master/arch/arm64/boot/dts/freescale/imx8mn-evk.dts#L33. For simplicity, I chose to have the DTS declare all PMIC registers, so that this driver could be made generic in the future.

mbolivar-nordic replied:
Thank you for the pointer! That indeed is more like what I was expecting -- there is a `compatible = "nxp,pca9450b";` in the Linux DTS, for instance.

DTS interoperability with Linux is a goal of Zephyr's (https://docs.zephyrproject.org/latest/guides/dts/design.html#source-compatibility-with-other-operating-systems), so a divergence here needs justification (https://docs.zephyrproject.org/latest/guides/dts/bindings.html#rules-for-mainline-bindings).

I'm not sure I follow the point about simplicity and making things generic, could you please clarify? I'm not too familiar with PMICs in general so sorry in advance for what might be a basic question.

mbolivar-nordic commented on drivers/regulator/regulator_i2c.c:
```diff
@@ -176,6 +246,7 @@ static const struct regulator_driver_api api = {
 #define CONFIGURE_REGULATOR(id)									\
 	static struct regulator_data i2c_reg_##id##_data;					\
 	static struct regulator_config i2c_reg_##id##_cfg = {					\
+		.label = DT_INST_PROP(id, label),						\
```
I don't think this is necessary. A devicetree label is used as the name field in a struct device by DEVICE_DT_INST_DEFINE already. See DEVICE_DT_NAME.

danieldegrasse replied:
Thanks for the heads up, I wasn't aware of the `name` field. I've removed the `label` field from the struct.

mbolivar-nordic commented on dts/bindings/regulator/regulator-i2c.yaml:
```diff
@@ -63,4 +63,30 @@ properties:
+        type: int
+        required: false
+        default: 0
+        description: Number of current limit levels this regulator supports
```
You must justify all default values in this binding; see https://docs.zephyrproject.org/latest/guides/dts/bindings.html#general-rules.

danieldegrasse replied:
I've added a short justification, essentially the driver uses the number of current limit levels to determine if the PMIC supports current limiting.

mbolivar-nordic commented on CODEOWNERS:
```diff
@@ -302,7 +302,7 @@
 /drivers/pwm/*gecko*                      @sun681
 /drivers/pwm/*it8xxx2*                    @RuibinChang
-/drivers/regulator/*i2c*		  @danieldegrasse
+/drivers/regulator/*pmic*		  @danieldegrasse
```
This PR has a setup based on the name "i2c", but then changes it later to use "pmic", here and elsewhere in this commit.

Please clean up the commits so the lines get added the way you want them the first time, without this type of change in a single PR.

danieldegrasse replied:
Fixed, thank you

danieldegrasse committed:
tests: regulator_pmic: Added PMIC regulator test

Test verifies that i2c pmic regulator can be disabled or enabled using
the regulator API. Verifies PMIC output voltage using ADC.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

danieldegrasse commented:
Updated the PR to include a test case for PMIC regulators, since this is a new driver

mbolivar-nordic commented:
Requested review from @ceolin and @JordanYates since this is touching regulators, and there are ongoing discussions about representing those in DT which may be relevant here

mbolivar-nordic commented:
> I'm not sure I follow the point about simplicity and making things generic, could you please clarify? I'm not too familiar with PMICs in general so sorry in advance for what might be a basic question.

I think this previous request from me is still open.

danieldegrasse commented:
> I think this previous request from me is still open.

Sorry about that! The way that linux handles PMIC regulators essentially requires each driver to provide a series of "register maps". The driver then uses a set of common support functions to implement PMIC control. For example, if you look at the [PCA9450 driver here](https://github.com/torvalds/linux/blob/master/drivers/regulator/pca9450-regulator.c#L73), you see that it uses functions like `regulator_enable_regmap`. The driver itself provides a series of register maps, which are defined using constants [here](https://github.com/torvalds/linux/blob/master/drivers/regulator/pca9450-regulator.c#L73).

The main reason I deviated from linux's implementation was that I didn't see the value in adding the additional register map layer. Almost all of the regulators in the linux tree are implemented using this `regmap` support structure, so I believe regulators could be ported to Zephyr without a need for a specific driver to declare registers, and those registers could be declared in the DTS.

The second change from the linux driver was how I declared regulator devices. Linux declares a regulator device based on the physical IC, and then the driver will expose individual regulators. For example, if there is a 3.3V regulator, and a 1.8V regulator on a single PMIC, linux will expose both within the same device. The user would then call [regulator_get()](https://www.kernel.org/doc/html/v4.15/driver-api/regulator.html#c.regulator_get), with the name of the regulator. I felt that the compute penalty required for looking up the regulator device was unnecessary, and that the existing regulator API would not fit well with this design (since you would need to get a reference to each individual regulator, then use the API to disable or enable it)

mbolivar-nordic reviewed (approved):
Thanks for justifying the divergence from Linux here. No objections from me, pinging @galak on the off chance he might be interested as well.
